### ISSUE TITLE
What should be paid attention to when training gpt-neox-20b with lora

### ISSUE THREAD

**Issue by kebijuelun** [lvwerra/trl#245]
https://github.com/lvwerra/trl/blob/main/examples/sentiment/scripts/gpt-neox-20b_peft/gpt-neo-20b_sentiment_peft.py#L181

The code mentioned that some hacky workaround need be done when training gpt-neox. Could you explain what is the problem when training gpt-neox-20b with lora.

**Comment by younesbelkada** [comment]
Hello @kebijuelun 
Thanks for the issue!
I believe the hack workaround is simply to not upcast the layernorm in fp32, thus keep them in fp16. 
Instead of calling the entire block: https://github.com/lvwerra/trl/blob/main/examples/sentiment/scripts/gpt-neox-20b_peft/gpt-neo-20b_sentiment_peft.py#L181-L190 one could have just add `layer_norm_names=[]` when calling `prepare_model_for_int8_training`
Maybe @edbeeching can confirm, but that is what I have got from my experience

**Comment by younesbelkada** [comment]
This is odd and should work, are you sure you are training your model in 8bit? 
https://github.com/kebijuelun/alpaca-lora-gpt/blob/c452b8dea635aa48c5c2e82e1bea25335ba3b9c4/finetune_gpt_neox_20b.py#L66
It seems here that you are loading the model in full precision, you need to pass `load_in_8bit=True` and `device_map="auto"`, hence the OOM I believe

**Comment by akk-123** [comment]
@edbeeching  great work! do you try training llama with lora?

**Comment by lvwerra** [comment]
Yes, we used it for StackLlama as well! See `examples/stack_llama/scripts`.

Closing this for now. Feel free to re-open if there open questions.
